core/unix/loader_linux.c

```diff
@@ -283,7 +284,98 @@ privload_mod_tls_init(privmod_t *mod)
 }
 
 static void
-privload_copy_tls_block(app_pc priv_tls_base, uint mod_idx)
+privload_set_pthread_tls_fields(privmod_t *mod, app_pc priv_tls_base)
+{
+    /* Set pthreads fields for glibc 3.24+ where pthreads is inside libc and has
+     * special undocumented initialization by ld.so.
+     */
+    if (strstr(mod->name, "libc.so") != mod->name)
+        return;
+    os_privmod_data_t *opd = (os_privmod_data_t *)mod->os_privmod_data;
+    /* i#6611: Find the pthread tid TLS field offset by decoding a function known
+     * to reference it.
+     */
+#    define PTHREAD_TID_FUNC_NAME "pthread_mutex_consistent"
+    void (*tid_using_func)(bool) = (void (*)(bool))get_proc_address_from_os_data(
+        &opd->os_data, opd->load_delta, PTHREAD_TID_FUNC_NAME, NULL);
+    if (tid_using_func == NULL)
+        return;
+    LOG(GLOBAL, LOG_LOADER, 2, "%s: decoding %s to find tid offset\n", __FUNCTION__,
+        PTHREAD_TID_FUNC_NAME);
+    app_pc pc = (app_pc)tid_using_func;
+    instr_t instr;
+    dcontext_t *dcontext = get_thread_private_dcontext();
+    if (dcontext == NULL)
+        dcontext = GLOBAL_DCONTEXT;
+    instr_init(dcontext, &instr);
+#    define MAX_INSTRS_TO_DECODE 64
+    int instr_count = 0;
+    long *tid_slot = NULL;
+    while (instr_count < MAX_INSTRS_TO_DECODE) {
+        IF_DEBUG(app_pc prev_pc = pc;)
+        pc = decode(dcontext, pc, &instr);
+        if (pc == NULL || !instr_valid(&instr)) {
+            SYSLOG_INTERNAL_WARNING("%s: failed to decode from %p\n", __FUNCTION__,
+                                    prev_pc);
+            break;
+        }
+        if (instr_is_return(&instr)) {
+            SYSLOG_INTERNAL_WARNING("%s: failed to find TLS offset\n", __FUNCTION__);
+            break;
+        }
+#    ifdef X86
+        /* We're looking for the only far ref in the function, like this:
+         *   8ac26:       64 8b 04 25 d0 02 00    mov    %fs:0x2d0,%eax
+         *   8ac2d:       00
+         */
+        if (instr_get_opcode(&instr) == OP_mov_ld &&
+            opnd_is_far_base_disp(instr_get_src(&instr, 0)) &&
+            opnd_get_segment(instr_get_src(&instr, 0)) == LIB_SEG_TLS) {
+            int offs = opnd_get_disp(instr_get_src(&instr, 0));
+            tid_slot = (long *)(priv_tls_base + offs);
+        }
+#    elif defined(AARCH64)
+        /* TODO i#6611: We have to decode something like this to come up with
+         * privlib_tls_base - 0x700 + 208 but we need a 2.37+ machine to test on:
+         *     13dec:       d53bd042        mrs     x2, tpidr_el0
+         *     13df0:       52800000        mov     w0, #0x0
+         *     13df4:       d11c0042        sub     x2, x2, #0x700
+         *     13df8:       b940d042        ldr     w2, [x2, #208]
+         */
+        break;
+#    else
+        /* XXX i#6611: Not supported. */
```

Comment:
We would presumably need to add this on RISC-V at some point right? s/supported/supported yet/?

Reply:
Yes.  Done.